GVFS/GVFS.Common/Tracing/JsonTracer.cs

```diff
@@ -339,40 +340,39 @@ private void WriteEvent(string eventName, EventLevel level, Keywords keywords, E
 
         private void MarkAndLogListenerRecovery(EventListener recoveredListener)
         {
+            // Check ContainsKey first (rather than always calling TryRemove) because ContainsKey
+            // is lock-free and recoveredListener should rarely be in failedListeners
             if (!this.failedListeners.ContainsKey(recoveredListener))
             {
-                // This listener has not failed since the last time it was called, so no need to do anything
+                // This listener has not failed since the last time it was called, so no need to log recovery
                 return;
             }
 
-            this.failedListeners.Remove(recoveredListener);
-
-            TraceEventMessage message = CreateListenerRecoveryMessage(recoveredListener);
-
-            // Only log that the listener has recovered to the other good listeners
-            foreach (EventListener listener in this.listeners.Except(this.failedListeners.Keys))
+            if (this.failedListeners.TryRemove(recoveredListener, out _))
             {
-                listener.TryRecordMessage(message, out _);
+                TraceEventMessage message = CreateListenerRecoveryMessage(recoveredListener);
+
+                // Only log that the listener has recovered to the other good listeners
+                foreach (EventListener listener in this.listeners.Except(this.failedListeners.Keys))
+                {
+                    listener.TryRecordMessage(message, out _);
+                }
             }
         }
 
         private void MarkAndLogListenerFailure(EventListener failedListener, string errorMessage)
         {
-            if (this.failedListeners.ContainsKey(failedListener))
+            // If we've already logged that this listener has failed there is no need to do it again

```

Comment:
I took the `ContainsKey` check out here because the failure scenario shouldn't be a hot path (and if it is we have bigger problems!)

Reply:
We could keep the previous structure (and avoid adding a level of nesting) by doing something like:

```
if (!this.failedListeners.TryAdd(….))
{
    return;
}
```